src/d/actor/d_a_kui.cpp

```diff
@@ -4,68 +4,541 @@
+
+        MtxTrans(actor->current.pos.x, actor->current.pos.y, actor->current.pos.z, FALSE);
+        mDoMtx_YrotM(*calc_mtx, actor->current.angle.y);
+        MtxScale(actor->scale.x, actor->scale.y, actor->scale.z, TRUE);
+        cMtx_copy(*calc_mtx, i_this->field_0x2A8);
+        i_this->field_0x2D8->Move();
+    } else {
+        MtxTrans(actor->current.pos.x, actor->current.pos.y, actor->current.pos.z, FALSE);
+        mDoMtx_YrotM(*calc_mtx, actor->current.angle.y);
+        if (i_this->type == 2 || i_this->type == 4) {
+            i_this->mpModel->setBaseTRMtx(*calc_mtx);
+        }
+
+        i_this->mpModel2->setBaseScale(actor->scale);
+        i_this->mpModel2->setBaseTRMtx(*calc_mtx);
+        if (i_this->field_0x2A2) {
+            cMtx_scale(local_mtx, 4.0f, 4.0f, 4.0f);
+        } else {
+            cMtx_scale(local_mtx, actor->scale.x, actor->scale.y, actor->scale.z);
+        }
+
+        cMtx_concat(*calc_mtx, local_mtx, i_this->field_0x2A8);
+        i_this->field_0x2D8->Move();
+    }
+
+    if (i_this->type == 2 || i_this->type == 4) {
+        BOOL is_switch = dComIfGs_isSwitch(i_this->mSwitchNo, dComIfGp_roomControl_getStayNo());
+
+        if (!is_switch) {
+            if (actor->health == 3 && i_this->field_0x308 == 0) {
+                i_this->field_0x308 = 1000;
+            }            
+        } else {
+            actor->current.pos.y = actor->home.pos.y - 70.0f;
+        }
+
+        if (i_this->field_0x308 != 0) {
+            i_this->field_0x308--;
+
+            s16 finished = REG8_S(3) + 0x3CA;
+            if (i_this->field_0x308 == finished) {
+                if (i_this->type == 2) {
+                    i_this->field_0x2E8 = 1;
+                } else {
+                    fopAcM_seStartCurrent(actor, JA_SE_OBJ_ROPE_SW_ON, 0);
+                    dComIfGs_onSwitch(i_this->mSwitchNo, fopAcM_GetRoomNo(actor));
+                    mDoAud_seStart(JA_SE_READ_RIDDLE_1);
+                }
+            }
+        }
+        demo_camera(i_this);
+    }
+
+    actor->eyePos = actor->current.pos;
+    return TRUE;
 }
 
 /* 000012E4-000012EC       .text daKui_IsDelete__FP9kui_class */
 static BOOL daKui_IsDelete(kui_class*) {
-    /* Nonmatching */
+    return TRUE;
 }
 
 /* 000012EC-00001340       .text daKui_Delete__FP9kui_class */
-static BOOL daKui_Delete(kui_class*) {
-    /* Nonmatching */
+static BOOL daKui_Delete(kui_class* i_this) {
+    dComIfG_resDelete(&i_this->mPhs, "Kui");
+    dComIfG_Bgsp()->Release(i_this->field_0x2D8);
+    return TRUE;
 }
 
 /* 00001340-00001664       .text daKui_CreateHeap__FP10fopAc_ac_c */
-static BOOL daKui_CreateHeap(fopAc_ac_c*) {
-    /* Nonmatching */
+static BOOL daKui_CreateHeap(fopAc_ac_c* a_this) {
+    kui_class* i_this = (kui_class*)a_this;
+    J3DModelData* modelData;
+
+    if (i_this->type == 3) {
+        // Bell body
+        modelData = (J3DModelData*)dComIfG_getObjectRes("Kui", KUI_BDL_HKANE1);
+        JUT_ASSERT(0x353, modelData != 0);
```

Comment:
Defines in asserts get replaced with their values, so these were probably written more like this originally
```suggestion
        JUT_ASSERT(0x353, modelData != NULL);
```